diamond-api/src/storage.rs

```diff
@@ -0,0 +1,217 @@
+use actix_web::error::Error;
+use glob::Pattern;
+use serde::*;
+use std::ffi::OsStr;
+use std::fs;
+use std::iter::successors;
+use std::path::{Path, PathBuf};
+use std::time::{SystemTime, UNIX_EPOCH};
+use whisper::interval::Interval;
+use whisper::{ArchiveData, WhisperFile};
+
+use crate::error::ResponseError;
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct RenderPoint(pub Option<f64>, pub u32);
+
+#[derive(Debug, Clone, PartialEq, PartialOrd, Serialize, Deserialize)]
+pub struct MetricResponseLeaf {
+    pub name: String,
+    pub path: String,
+    pub is_leaf: bool,
+}
+
+pub trait Walker {
+    fn walk(&self, metric: &str, interval: Interval) -> Result<Vec<RenderPoint>, ResponseError>;
+    fn walk_tree(&self, subdir: &Path, pattern: &Pattern)
+        -> Result<Vec<MetricResponseLeaf>, Error>;
+}
+
+#[derive(Clone)]
+pub struct WalkerPath(pub PathBuf);
+
+impl Walker for WalkerPath {
+    fn walk(&self, metric: &str, interval: Interval) -> Result<Vec<RenderPoint>, ResponseError> {
+        let full_path = path(self.0.as_path(), metric)?;
+        let now = SystemTime::now().duration_since(UNIX_EPOCH)?.as_secs() as u32;
+        let ArchiveData {
+            from_interval,
+            step,
+            values,
+            ..
+        } = WhisperFile::open(&full_path)?.fetch_auto_points(interval, now)?;
+        let timestamps = successors(Some(from_interval), |i| i.checked_add(step));
+        let points = values
+            .into_iter()
+            .zip(timestamps)
+            .map(|(value, time)| RenderPoint(value, time))
+            .collect();
+
+        Ok(points)
+    }
+
+    fn walk_tree(
+        &self,
+        subdir: &Path,
+        pattern: &Pattern,
+    ) -> Result<Vec<MetricResponseLeaf>, Error> {
+        let full_path = self.0.canonicalize()?.join(&subdir);
+        let dir_metric = subdir
+            .components()
+            .filter_map(|x| x.as_os_str().to_str())
+            .collect::<Vec<&str>>()
+            .concat();
```

Comment:
I have a feeling that `fold` will look better here. Similar to what you did with `PathBuf` below but with a `String`.